pkg/appgateserver/session.go

```diff
@@ -0,0 +1,47 @@
+package appgateserver
+
+import (
+	"context"
+
+	sessiontypes "github.com/pokt-network/poktroll/x/session/types"
+	sharedtypes "github.com/pokt-network/poktroll/x/shared/types"
+)
+
+// getCurrentSession gets the current session for the given service
+// It returns the current session if it exists and is still valid, otherwise it
+// queries for the latest session, caches and returns it.
+func (app *appGateServer) getCurrentSession(
+	ctx context.Context,
+	appAddress, serviceId string,
+) (*sessiontypes.Session, error) {
+	app.sessionMu.RLock()
+	defer app.sessionMu.RUnlock()
+
+	latestBlock := app.blockClient.LatestBlock(ctx)
+	if currentSession, ok := app.currentSessions[serviceId]; ok {
+		sessionEndBlockHeight := currentSession.Header.SessionStartBlockHeight + currentSession.NumBlocksPerSession
+
+		// Return the current session if it is still valid.
```

Comment:
DO we want to return an error in the else case?

Reply:
No the else case is querying the chain for the current session and updating the cache